File: Medii de Proiectare si Programare/HW/mpp_hw/src/main/java/mpp/com/Service/Service.java
package mpp.com.Service;

import mpp.com.Domain.Child;
import mpp.com.Domain.ChildTrackDTO;
import mpp.com.Domain.Track;
import mpp.com.Domain.User;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class Service {

    ServiceUser serviceUser;
    ServiceTrack serviceTrack;
    ServiceChild serviceChild;

    public Service(ServiceUser serviceUser, ServiceTrack serviceTrack, ServiceChild serviceChild) {
        this.serviceUser = serviceUser;
        this.serviceTrack = serviceTrack;
        this.serviceChild = serviceChild;
    }

    public Optional<User> searchUsedByUsernamePassword(String username, String password) throws IllegalArgumentException {
        return serviceUser.searchUsedByUsernamePassword(username, password);
    }

    public Optional<Track> searchTrackById(Long idTrack) throws IllegalArgumentException {
        return serviceTrack.searchTrackById(idTrack);
    }

    public Iterable<Track> getAllTracks() {
        return serviceTrack.getAllTracks();
    }

    public void addTrack(String name, int minimumAge, int maximumAge, int distance) {
        serviceTrack.addTrack(name, minimumAge, maximumAge, distance);
    }

    public Optional<Child> searchChildById(Long id) throws IllegalArgumentException {
        return serviceChild.searchChildById(id);
    }

    public Iterable<Child> getAllChildren() {
        return serviceChild.getAllChildren();
    }

    public void addChild(String name, int age, String[] trackValues) {
        List<Track> tracks = new ArrayList<>();
        for(String trackName : trackValues) {
            try{
                var track = serviceTrack.findTrackByName(trackName);
                if (track.get().getMinimumAge() > age || track.get().getMaximumAge() < age)
                    throw new RuntimeException("The child's age is not in the track's age range");
                tracks.add(track.get());
            } catch (Exception e) {
                throw new RuntimeException(e.getMessage());
            }
        }
        serviceChild.addChild(name, age,tracks);
    }

    public Iterable<ChildTrackDTO> getChildrenByTracks(Long idTrack) {
        return serviceChild.getChildrenByTracks(idTrack);
    }

    public Optional<Track> findTrackByName(String trackName) {
        return serviceTrack.findTrackByName(trackName);
    }

}

File: Medii de Proiectare si Programare/HW/MPP_CSharp/MPP_CSharp/Service/Service.cs
using System;
using System.Collections.Generic;
using MPP_CSharp.Domain;
using MPP_CSharp.Service;

namespace MPP_CSharp.Service
{
    public class Service
    {
        private readonly ServiceUser _serviceUser;
        private readonly ServiceTrack _serviceTrack;
        private readonly ServiceChild _serviceChild;

        public Service(ServiceUser serviceUser, ServiceTrack serviceTrack, ServiceChild serviceChild)
        {
            _serviceUser = serviceUser ?? throw new ArgumentNullException(nameof(serviceUser));
            _serviceTrack = serviceTrack ?? throw new ArgumentNullException(nameof(serviceTrack));
            _serviceChild = serviceChild ?? throw new ArgumentNullException(nameof(serviceChild));
        }

        public User? SearchUserByUsernamePassword(string username, string password)
        {
            return _serviceUser.SearchUserByUsernamePassword(username, password);
        }

        public Track? SearchTrackById(long idTrack)
        {
            return _serviceTrack.SearchTrackById(idTrack);
        }

        public IEnumerable<Track> GetAllTracks()
        {
            return _serviceTrack.GetAllTracks();
        }

        public void AddTrack(string name, int minimumAge, int maximumAge, int distance)
        {
            _serviceTrack.AddTrack(name, minimumAge, maximumAge, distance);
        }

        public Child? SearchChildById(long id)
        {
            return _serviceChild.SearchChildById(id);
        }

        public IEnumerable<Child> GetAllChildren()
        {
            return _serviceChild.GetAllChildren();
        }

        public void AddChild(string name, int age, string[] trackValues)
        {
            var tracks = new List<Track>();
            foreach (var trackName in trackValues)
            {
                var track = _serviceTrack.FindTrackByName(trackName);
                if (track == null || track.MinimumAge > age || track.MaximumAge < age)
                {
                    throw new Exception("The child's age is not in the track's age range");
                }
                tracks.Add(track);
            }
            _serviceChild.AddChild(name, age, tracks);
        }

        public IEnumerable<ChildTrackDTO> GetChildrenByTracks(long idTrack)
        {
            return _serviceChild.GetChildrenByTracks(idTrack);
        }

        public Track? FindTrackByName(string trackName)
        {
            return _serviceTrack.FindTrackByName(trackName);
        }

        public IEnumerable<Track> GetTracksByAge(int minimumAge, int maximumAge)
        {
            return _serviceTrack.GetTracksByAge(minimumAge, maximumAge);
        }

        public IEnumerable<Child> GetChildrenByTrackId(long idTrack)
        {
            return _serviceChild.GetChildrenByTrackId(idTrack);
        }

        public IEnumerable<TrackDTO> GetTrackDTOs()
        {
            return _serviceTrack.GetTrackDTOs();
        }

        public IEnumerable<TrackDTO> GetTrackDTOsByAge(int minimumAge, int maximumAge)
        {
            return _serviceTrack.GetTrackDTOsByAge(minimumAge, maximumAge);
        }
    }
}